File: app/src/main/java/io/supertokens/session/utils/AntiCSRF.java
package io.supertokens.session.utils;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;
import androidx.annotation.Nullable;
import io.supertokens.session.R;

import java.util.HashMap;

@SuppressWarnings("Convert2Diamond")
public class AntiCSRF {
    private static HashMap<String, String> antiCSRFTokenInfo;
    private static final  Object lock = new Object();
    private static final String antiCSRFMapKey = "antiCSRF";
    private static final String associatedIdRefreshMapKey = "associatedIdRefreshToken";

    @SuppressWarnings("ConstantConditions")
    public static String getToken(Context context, @Nullable String associatedRefreshToken) {
        synchronized (lock) {
            if ( associatedRefreshToken == null ) {
                return null;
            }

            if ( antiCSRFTokenInfo == null ) {
                SharedPreferences sharedPreferences = getSharedPreferences(context);
                String antiCSRF = sharedPreferences.getString(getSharedPrefsAntiCSRFKey(context), null);
                if ( antiCSRF == null ) {
                    return null;
                }

                antiCSRFTokenInfo = new HashMap<String, String>();
                antiCSRFTokenInfo.put(antiCSRFMapKey, antiCSRF);
                antiCSRFTokenInfo.put(associatedIdRefreshMapKey, associatedRefreshToken);
            } else if ( antiCSRFTokenInfo.get(associatedIdRefreshMapKey) != null && !antiCSRFTokenInfo.get(associatedIdRefreshMapKey).equals(associatedRefreshToken) ) {
                antiCSRFTokenInfo = null;
                return getToken(context, associatedRefreshToken);
            }

            return antiCSRFTokenInfo.get(antiCSRFMapKey);
        }
    }

    @SuppressLint("ApplySharedPref")
    public static void removeToken(Context context) {
        synchronized (lock) {
            SharedPreferences sharedPreferences = getSharedPreferences(context);
            SharedPreferences.Editor editor = sharedPreferences.edit();
            editor.remove(getSharedPrefsAntiCSRFKey(context));
            editor.commit();
        }
    }

    @SuppressLint("ApplySharedPref")
    public static void setToken(Context context, @Nullable String associatedRefreshToken, String antiCSRFToken) {
        synchronized (lock) {
            if ( associatedRefreshToken == null ) {
                antiCSRFTokenInfo = null;
                return;
            }

            SharedPreferences sharedPreferences = getSharedPreferences(context);
            SharedPreferences.Editor editor = sharedPreferences.edit();
            editor.putString(getSharedPrefsAntiCSRFKey(context), antiCSRFToken);
            editor.commit();

            antiCSRFTokenInfo = new HashMap<String, String>();
            antiCSRFTokenInfo.put(antiCSRFMapKey, antiCSRFToken);
            antiCSRFTokenInfo.put(associatedIdRefreshMapKey, associatedRefreshToken);
        }
    }

    private static SharedPreferences getSharedPreferences(Context context) {
        return context.getSharedPreferences(context.getString(R.string.supertokensSharedPrefsKey), Context.MODE_PRIVATE);
    }

    private static String getSharedPrefsAntiCSRFKey(Context context) {
        return context.getString(R.string.supertokensAntiCSRFTokenKey);
    }
}

File: app/src/main/java/io/supertokens/session/utils/AntiCSRF.kt
package io.supertokens.session.utils

import android.content.Context
import android.content.SharedPreferences
import io.supertokens.session.R

class AntiCSRF {
    companion object {
        @JvmStatic
        private var anticsrftokenInfo: MutableMap<String, String>? = null

        @JvmStatic
        private val lock = Object()

        @JvmStatic
        fun getToken(context: Context, associatedIdRefreshToken: String?): String? = synchronized(lock) {
            if ( associatedIdRefreshToken == null ) {
                anticsrftokenInfo = null
                return null
            }
            if ( anticsrftokenInfo == null ) {
                val sharedPreferences = getSharedPreferences(context)
                val anticsrf = sharedPreferences.getString(getSharedPrefsAntiCSRFKey(context), null)
                if ( anticsrf == null ) {
                    return null
                }
                anticsrftokenInfo = mutableMapOf()
                anticsrftokenInfo!!["antiCSRF"] = anticsrf
                anticsrftokenInfo!!["associatedIdRefreshToken"] = associatedIdRefreshToken
            } else if(anticsrftokenInfo!!["associatedIdRefreshToken"] != associatedIdRefreshToken) {
                anticsrftokenInfo = null
                return getToken(context, associatedIdRefreshToken)
            }
            return anticsrftokenInfo!!["antiCSRF"]
        }

        @JvmStatic
        fun removeToken(context: Context) {
            synchronized(lock) {
                val sharedPreferences = getSharedPreferences(context)
                with(sharedPreferences.edit()) {
                    remove(getSharedPrefsAntiCSRFKey(context))
                    commit()
                }
            }
        }

        @JvmStatic
        fun setToken(context: Context, associatedIdRefreshToken: String?, antiCSRF: String) {
            synchronized(lock) {
                if ( associatedIdRefreshToken == null ) {
                    anticsrftokenInfo = null
                    return
                }
                val sharedPreferences = getSharedPreferences(context)
                with(sharedPreferences.edit()) {
                    putString(getSharedPrefsAntiCSRFKey(context), antiCSRF)
                    commit()
                }
                anticsrftokenInfo = mutableMapOf()
                anticsrftokenInfo!!["antiCSRF"] = antiCSRF
                anticsrftokenInfo!!["associatedIdRefreshToken"] = associatedIdRefreshToken
            }
        }

        @JvmStatic
        private fun getSharedPreferences(context: Context): SharedPreferences =
            context.getSharedPreferences(context.getString(R.string.supertokensSharedPrefsKey), Context.MODE_PRIVATE)

        @JvmStatic
        private fun getSharedPrefsAntiCSRFKey(context: Context): String = context.getString(R.string.supertokensAntiCSRFTokenKey)
    }
}